core/src/test/java/com/linecorp/armeria/server/decoding/DecodingServiceTest.java

```diff
@@ -0,0 +1,100 @@
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.decoding;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.zip.DeflaterOutputStream;
+import java.util.zip.GZIPOutputStream;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import com.linecorp.armeria.client.WebClient;
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.MediaType;
+import com.linecorp.armeria.common.RequestHeaders;
+import com.linecorp.armeria.common.ResponseHeaders;
+import com.linecorp.armeria.server.HttpService;
+import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.testing.junit5.server.ServerExtension;
+
+class DecodingServiceTest {
+
+    @RegisterExtension
+    static final ServerExtension server = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) throws Exception {
+            final HttpService httpService = (ctx, req) -> HttpResponse.from(
+                    req.aggregate()
+                       .thenApply(aggregated -> {
+                           final ResponseHeaders responseHeaders =
+                                   ResponseHeaders.of(HttpStatus.OK,
+                                                      HttpHeaderNames.CONTENT_TYPE, MediaType.PLAIN_TEXT_UTF_8);
+                           return HttpResponse.of(responseHeaders,
+                                                  HttpData.ofUtf8("Hello " + aggregated.contentUtf8() + '!'));
+                       }));
+            sb.decorator("/decodeTest", DecodingService.newDecorator());
+            sb.service("/decodeTest", httpService);
+
+            sb.decoratorUnder("/", (delegate, ctx, req) -> {
+                return delegate.serve(ctx, req);
+            });
+        }
+    };
+
+    @Test
+    void decodingGzippedPayloadFromClient() throws IOException {
+
+        final WebClient client = WebClient.builder(server.httpUri()).build();
+        final RequestHeaders headers = RequestHeaders.of(HttpMethod.POST, "/decodeTest",
+                                                         HttpHeaderNames.CONTENT_ENCODING, "gzip");
+        final ByteArrayOutputStream encodedStream = new ByteArrayOutputStream();
+        final DeflaterOutputStream encodingStream = new GZIPOutputStream(encodedStream, true);
+
+        final byte[] testByteArray = "Armeria Gzip Test".getBytes(StandardCharsets.UTF_8);
+        encodingStream.write(testByteArray);
+        encodingStream.flush();
+
+        assertThat(client.execute(headers, HttpData.wrap(encodedStream.toByteArray())).aggregate().join()
```

Comment:
And how about adding `EncodingClient` for completing flow - compressing by client and decompressing by server?
You can do that this PR or in the following PR. 😉

Reply:
`EncodingClient` is good idea that armeria support end-to-end compressed data for reduce traffic size. I'll do another PR to implement `EncodingClient` ASAP :smile:

Reply:
Awesome, @joonhaeng. Looking forward to the follow-up PR. ❤️